settings: Add options to automatically follow and/or unmute topics.

prakhar1144 opened the pull request:
Add option to:
- [x] Automatically Follow topics
- [x] Automatically Unmute topics
- [x] Automatically Follow topics (Frontend)
- [x] Automatically Unmute topics (Frontend)

Note:
* There is an increase in the query count by  5 (max) while sending a message.
  * When we send the first message to a topic, there is an increase in the query count by 5 --  1 for checking if it is the first message in the topic + 1 to check if the topic is already followed + 3 for following the topic.
 
Fixes #25914 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->
**Demo Video**

[automatic_3.webm](https://github.com/zulip/zulip/assets/56781761/8e368a95-2936-4b26-b807-34b9d9243156)




<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
Question:

If a user has selected the option: **Topics I participate in**

So, every time the user sends a message, `do_set_user_topic_visibility_policy` will try to follow the topic. If the topic is already followed, we will have a lot of INFO logs. for example:
> 2023-06-19 04:37:05.207 INFO [] User 10 tried to set visibility_policy to its current value of 3

We can avoid this with an initial check; if the topic is already followed, then skip. But this will cost us an extra query in the message send code path.

Suggestions on how to tackle this?

(check the changes in `zerver/actions/message_send.py` for related code)

abhijeetbodas2001 commented on zerver/tests/test_user_topics.py:
```diff
@@ -641,3 +643,189 @@ def test_unmuted_topic_remove_invalid(self) -> None:
+
+        # Cordelia reacts to a message. DO automatically follow the topic.
+        # Aaron reacts to a message. DON'T automatically follow the topic.
+        def add_reaction(user: UserProfile) -> None:
+            reaction_info = {
+                "emoji_name": "smile",
+            }
+            endpoint = f"/api/v1/messages/{message_id}/reactions"
+            result = self.api_post(user, endpoint, reaction_info)
+            self.assert_json_success(result)
+
+        add_reaction(cordelia)
+        add_reaction(aaron)
```
Might be more convenient to use `/actions/reactions.py/check_add_reaction` than calling the endpoint here.

prakhar1144 replied:
Used `check_add_reaction`

abhijeetbodas2001 reviewed (commented):
Thanks for the PR! This is a lot of progress. I left a review below. Some other comments:

* We will be implementing "Automatically unmute topics" later on as well. So if there are any easy-to-create abstractions that might help us later (like making generic functions etc), it would make sense do that. I don't have any specific suggestions right now, but please don't go overboard with this! We can always have more prep commits whenever we implement "Automatically unmute topics" if that makes it easier.
One more question is : would be easier (in terms of commit structure etc) to implement the "automatically unmute" part **in parallel** with this? What do you think?
* Regarding your question in https://github.com/zulip/zulip/pull/26054#issuecomment-1596487707, I think we can introduce a new parameter to the function called `silent_fail_for_duplicates` or something like that to avoid the problem of extra logging. But I feel we can easily implement that later if we decide to, once we have more feedback on this (feel free to raise this up on #backend).

alexmv commented on version.py:
```diff
@@ -33,7 +33,7 @@
 # entries in the endpoint's documentation in `zulip.yaml`.
-API_FEATURE_LEVEL = 188
+API_FEATURE_LEVEL = 190
 
```
Fixed in#26014.

alexmv commented on zerver/actions/message_send.py:
```diff
@@ -851,6 +853,40 @@ def do_send_messages(
+
+            # The sender automatically follows the topic if the sender has started the topic and
+            # 'automatically_follow_topics_policy' set to 'AUTOMATICALLY_FOLLOW_TOPICS_I_START'.
+            topic_messages = Message.objects.filter(
+                recipient=send_request.message.recipient,
+                subject__iexact=send_request.message.topic_name(),
+            )
```
We should also make sure we add a test that covers this specific case, because it could have security implications, and it's a bit fiddly.

prakhar1144 replied:
Done

alexmv commented on zerver/tests/test_message_edit.py:
```diff
@@ -1749,6 +1785,18 @@ def assert_has_visibility_policy(
         self.subscribe(aaron, stream_name)
         self.login_user(aaron)
 
+        # A user automatically follows a topic he/she starts by default.
```
```suggestion
        # A user automatically follows a topic they start by default.
```

alexmv commented on zerver/tests/test_message_edit.py:
```diff
@@ -1905,6 +1953,18 @@ def test_user_topic_states_on_moving_to_topic_with_no_messages(self) -> None:
         self.subscribe(cordelia, stream_name)
         self.subscribe(aaron, stream_name)
 
+        # A user automatically follows a topic he/she starts by default.
```
```suggestion
        # A user automatically follows a topic they start by default.
```

alexmv commented on zerver/tests/test_message_send.py:
```diff
@@ -1540,15 +1541,29 @@ def test_is_private_flag(self) -> None:
+    def _send_stream_message(
+        self, user: UserProfile, stream_name: str, content: str, expected_num_events: int = 1
+    ) -> Set[int]:
+        # Because a user automatically follows the topic he/she starts (by default),
```
```suggestion
        # Because a user automatically follows the topic they start (by default),
```

alexmv commented on zerver/tests/test_events.py:
```diff
@@ -516,12 +516,21 @@ def test_huddle_send_message_events(self) -> None:
         )
 
     def test_stream_send_message_events(self) -> None:
+        # Because a user automatically follows the topic he/she starts (by default),
```
```suggestion
       # Because a user automatically follows the topic they start (by default),
```

alexmv commented on zerver/tests/test_events.py:
```diff
@@ -2605,6 +2651,15 @@ def do_upload() -> None:
         self.subscribe(hamlet, "Denmark")
         assert url is not None
         body = f"First message ...[zulip.txt](http://{hamlet.realm.host}" + url + ")"
+        # A user automatically follows the topic he/she starts (by default).
```
```suggestion
        # A user automatically follows the topic they start (by default).
```

alexmv commented on zerver/tests/test_message_edit.py:
```diff
@@ -115,6 +116,17 @@ def prepare_move_topics(
         self.subscribe(user_profile, stream.name)
         self.subscribe(user_profile, stream_to.name)
+
+        # A user automatically follows a topic he/she starts by default.
```
```suggestion
        # A user automatically follows a topic they start by default.
```

alexmv commented on zerver/tests/test_message_edit.py:
```diff
@@ -1315,6 +1327,18 @@ def test_edit_muted_topic(self, mock_send_event: mock.MagicMock) -> None:
         self.subscribe(hamlet, stream_name)
         self.login_user(hamlet)
+
+        # A user automatically follows a topic he/she starts by default.
```
```suggestion
        # A user automatically follows a topic they start by default.
```

alexmv reviewed (commented):
Just adding a couple comments on very quick skim.  This is shaping up well!

prakhar1144 commented on zerver/tests/test_message_edit.py:
```diff
@@ -1749,6 +1785,18 @@ def assert_has_visibility_policy(
         self.subscribe(aaron, stream_name)
         self.login_user(aaron)
 
+        # A user automatically follows a topic he/she starts by default.
+        # We change the default behaviour at the start of this test,
+        # as our main focus is to test all possible combinations
+        # of orig_topic and target_topic visibility policies.
+        for user in [hamlet, cordelia, aaron]:
+            do_change_user_setting(
+                user_profile=user,
+                setting_name="automatically_follow_topics_policy",
+                setting_value=UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_NEVER,
+                acting_user=None,
+            )
+
```
[CZO Discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1600092)

prakhar1144 commented on zerver/tests/test_user_topics.py:
```diff
@@ -641,3 +643,189 @@ def test_unmuted_topic_remove_invalid(self) -> None:
+        self.assertEqual(user_ids, {hamlet.id})
+
+    def test_automatically_follow_topic_on_participation(self) -> None:
+        hamlet = self.example_user("hamlet")
```
Created separate tests

prakhar1144 commented on zerver/tests/test_user_topics.py:
```diff
@@ -641,3 +643,189 @@ def test_unmuted_topic_remove_invalid(self) -> None:
+        )
+        self.send_stream_message(iago, stream_name=stream.name, topic_name="New Topic")
+        user_ids = stream_topic_target.user_ids_with_visibility_policy(
+            UserTopic.VisibilityPolicy.FOLLOWED
```
Hm, right -- Fixed

prakhar1144 commented on zerver/tests/test_events.py:
```diff
@@ -516,12 +516,21 @@ def test_huddle_send_message_events(self) -> None:
+        # event-type: muted_topics (deprecated)
+        self.assertEqual(events[0]["muted_topics"], [])
+        # event-type: user_topic
+        self.assertEqual(events[1]["visibility_policy"], UserTopic.VisibilityPolicy.FOLLOWED)
```
Done

prakhar1144 commented on zerver/models.py:
```diff
@@ -1634,6 +1634,18 @@ class UserBaseSettings(models.Model):
+    AUTOMATICALLY_FOLLOW_TOPICS_POLICY_CHOICES = [
+        AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION,
+        AUTOMATICALLY_FOLLOW_TOPICS_I_START,
+        AUTOMATICALLY_FOLLOW_TOPICS_NEVER,
```
Changed to `ON_INITIATION`

prakhar1144 commented on zerver/actions/reactions.py:
```diff
@@ -85,6 +86,25 @@ def do_add_reaction(
+    ):
+        stream_id = message.recipient.type_id
+        stream = Stream.objects.select_related().get(id=stream_id)
+        assert stream is not None
```
Done

prakhar1144 commented on zerver/openapi/zulip.yaml:
```diff
@@ -11983,6 +12000,16 @@ paths:
+
+                              - 1 - Topics I participate in
+                              - 2 - Topics I start
+                              - 3 - Never
```
Hm, better -- made the changes

prakhar1144 commented:
@zulipbot add "mentor review"

timabbott commented:
Is there anything blocking integration of this other than a review from me?

prakhar1144 commented:
@timabbott 
Yes, pending discussion: https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1614384

prakhar1144 commented on zerver/tests/test_user_topics.py:
```diff
@@ -641,3 +643,189 @@ def test_unmuted_topic_remove_invalid(self) -> None:
+
+class AutomaticallyFollowTopicsTests(ZulipTestCase):
+    def test_automatically_follow_topic_you_start(self) -> None:
+        hamlet = self.example_user("hamlet")
```
Done

prakhar1144 commented:
@zulipbot add "QA"

prakhar1144 commented:
@zulipbot add "QA needed"

prakhar1144 commented:
@abhijeetbodas2001 @timabbott 

This is ready to review.

Updates in the recent push:
* Addressed the review comments.
* Added last 3 commits -- Automatically unmute policy + UI for these two settings.
* Resolved the error: [CZO discussion](https://chat.zulip.org/#narrow/stream/49-development-help/topic/test_custom_email_with_dry_run/near/1628433) -- first commit

abhijeetbodas2001 commented on zerver/actions/message_send.py:
```diff
@@ -913,6 +915,41 @@ def do_send_messages(
+                user_profile=sender,
+                message__recipient_id=send_request.message.recipient_id,
+                message__subject__iexact=send_request.message.topic_name(),
+            )
```
We can avoid the query if the policy isn't `ON_PARTICIPATION`, right? Currently we are running this query for all messages.
An `if-elif` sequence would be better here:
```
if not bot:
    if policy = ON_PARTICIPATION:
        # ...
    elif policy = ON_INITIATION:
        accessible_messages_in_topic = ...
        if len(accessible_messages_in_topic) > 0:
            # ...
```

abhijeetbodas2001 reviewed (commented):
Thanks for the update!
I read through all the code except the test in the automatically_unmute commit and the UI changes.
I left some comments on ways to deduplicate the tests and the visibility_policy calculation logic.

Two broader questions:
1. For testing scenarios when both the automatically_{follow,unmute} settings are involved, do we need to test the events every time? Why won't it suffice to write the tests in `test_user_topics.py` itself? That way, we can test that the topic is actually followed/unmuted, instead of just checking the number of events.
2. The `AutomaticallyUnmuteTopicsTests` and `AutomaticallyFollowTopicsTests` are very similar structurally, right? Can we deduplicate those?

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -913,6 +915,41 @@ def do_send_messages(
+                user_profile=sender,
+                message__recipient_id=send_request.message.recipient_id,
+                message__subject__iexact=send_request.message.topic_name(),
+            )
```
I have squashed the 2 commits introducing `follow` and `unmute` settings separately. Keeping them separate gives incomplete information, hence the above comment. `if policy = ON_PARTICIPATION:` - this kind of thing is not relevant when we have both settings involved.

In short: This comment resulted due to a overly-minimal commit structure. The comment can be ignored.

prakhar1144 commented on zerver/tests/test_events.py:
```diff
@@ -2692,6 +2712,19 @@ def do_upload() -> None:
+            "automatically_follow_topics_policy",
+            UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_NEVER,
+            acting_user=None,
+        )
```
> I think we had decided to set the policy for test users to be NEVER in the test database? 

Yes, I finally did that.

Earlier, I set the policy in the individual tests themselves because, as [Steve said](https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1602472):
> Generally I have found it easier when writing (and reading) tests to just explicitly set up the data in individual tests. The set_up() methods can be kind of tricky for future developers who don't realize right away that there is action-at-a-distance.

But I agree that it is too verbose. So [Tim's suggestion](https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1609293) would be the better approach here:
>  we can set a defaults during the creation of test users in populate_db, though.

prakhar1144 commented on zerver/actions/message_send.py:
```diff
@@ -917,39 +919,94 @@ def do_send_messages(
+            #       ON_PARTICIPATION               |           ON_INITIATION              |     FOLLOWED
+            #         ON_INITIATION                |          ON_PARTICIPATION            |     FOLLOWED
+            #       ON_PARTICIPATION               |          ON_PARTICIPATION            |     FOLLOWED
+            #         ON_INITIATION                |           ON_INITIATION              |     FOLLOWED
```
Improved comment -- [see lib/message.py](https://github.com/zulip/zulip/pull/26054/files#diff-40c1c91d1032197fa2decf32d350e72180d5861534c2eb99fcc898eb0b12da24R1737)

prakhar1144 commented:
This is ready to review

Updates in the recent push:
* Addressed the review comments (check respective comments)
*  we have UI and API level constraints to disallow combinations of settings like **both having the same value, except NEVER** -- [related CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1629977)

timabbott commented:
I merged the prep commits as the series ending with db8229ae32475ec9d114d04a35a50c5096d27f2e.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -917,39 +919,94 @@ def do_send_messages(
+                sender.automatically_follow_topics_policy
                 == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION
+                or sender.automatically_unmute_topics_policy
+                == UserProfile.AUTOMATICALLY_UNMUTE_TOPICS_ON_PARTICIPATION
```
Yeah I like these interfaces!

timabbott commented on zerver/lib/message.py:
```diff
@@ -1697,3 +1697,109 @@ def update_to_dict_cache(
+    )
+    if not user_topic:
+        return True
+    current_visibility_policy = user_topic[0].visibility_policy
```
I would usually write this using `.get` and just catching the `UserTopic.DoesNotExist` exception to do the `return True` case.

prakhar1144 replied:
Done.

timabbott commented on zerver/lib/message.py:
```diff
@@ -1697,3 +1697,109 @@ def update_to_dict_cache(
+        and accessible_messages_in_topic.count() == 1
+    ):
+        return UserTopic.VisibilityPolicy.FOLLOWED
+
```
I think it might be worth adding an `if visibility_policy == UserTopic.VisibilityPolicy.UNMUTED: return ...` just after this -- the query for how many messages there are can be expensive, and so we should avoid doing it when not necessary.

prakhar1144 replied:
We're now using `.exclude(id=message.id).exists()`; 

It will be called only when:
* automatically_follow_topics_policy == AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
* automatically_unmute_topics_policy == AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION

overall one db query when either or both of the policy is `ON_INITIATION`.

and it is necessary to check if it is the topic start when the policy is `ON_INITIATION`

So, we can avoid adding `if visibility_policy == UserTopic.VisibilityPolicy.UNMUTED: return ...`

timabbott replied:
I think we also want to avoid doing that calculation if we know that the user already has the state that this would trigger. Otherwise, if you send a message to a topic 10 times in a row, we can potentially be doing this query for 9/10 of them for no reason.

prakhar1144 replied:
I think things would be much easier in future when we'll have a `Topic` or `StreamTopic` table. We can simply use a `is_new_topic = True` variable parallely while creating a new `topic` object in the send message codepath; that would lead to avoiding this unnecessart repetitive query.

For now, I'm not sure what would be the ideal place to store the info "is_new_topic = False" after a user finishes sending the first message and reuse during next send message operation.

>  if we know that the user already has the state that this would trigger.

The problem would be for example: If a user has
* follow_policy = on_initiation
* unmute_policy = never

Now if the user sends a message to an existing topic & they have the no visibility policy set; the db query will occur even if we apply the check "if we know that the user already has the state that this would trigger. "

prakhar1144 replied:
@timabbott 
ICYMI: we have one more question to discuss!

timabbott replied:
Regarding a topic table I think we won't be able to do an `is_new_topic` check that way for private streams with protected history (since we can't leak to the user whether there are previous messages); but those are rare, and in shared history streams, it will be a nice optimization. Just wanted to mention that's a corner case that a topics table doesn't make any less ugly.

> Now if the user sends a message to an existing topic & they have the no visibility policy set; the db query will occur even if we apply the check "if we know that the user already has the state that this would trigger. "

I think the fix for this is that we need to pass `current_visibility_policy` to this function, and we can skip fetching existing messages if the user's current_visibility_policy for the topic is already at least as visible as what the policy would change it to.

prakhar1144 replied:
A better fact that we can use to solve this is:

"If a topic has a visibility policy set, it can't be the case of initiation. We return early, thus saving a DB query."

We now pass `current_visibility_policy` to this function and return None if it can't increase visibility.

`should_change_visibility_policy` is now only used in reactions and submessage codepath.

timabbott replied:
OK, this is pretty good! This should probably be a follow-up for sanity's sake, but we might want to make sure we have tests for the query counts for adding a reaction / etc., maybe in functions like ` test_automatically_unmute_topic_on_participation_add_reaction`?

timabbott commented on zerver/lib/message.py:
```diff
@@ -1697,3 +1697,109 @@ def update_to_dict_cache(
+        user_profile=sender,
+        message__recipient_id=message.recipient_id,
+        message__subject__iexact=message.topic_name(),
+    )
```
This query is probably more expensive than would be ideal -- if there's 800 messsages in the topic, we end up asking the database to count them all.

In theory all we need is `.exclude(id=message.id).exists()` for the below calculations.

prakhar1144 replied:
Yes, using `.exclude(id=message.id).exists()` is far better. Done.

prakhar1144 commented:
@timabbott This is ready to review.

Updates in the recent push:
* Addressed the review comments.

The test failure is related to [`postgres client upgrade failures`](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/postgres.20client.20upgrade.20failures/near/1640444)

timabbott commented on zerver/lib/message.py:
```diff
@@ -1697,3 +1697,110 @@ def update_to_dict_cache(
+
+    if (
+        sender.automatically_unmute_topics_policy
+        == UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION
```
This should be `AUTOMATICALLY_UNMUTE_TOPICS_ON_PARTICIPATION`, right?

I think we might want to just use a common set of values, and name them e.g. 
`CHANGE_USER_TOPIC_ON_INITIATION`, just to avoid the possibility of mixing this up and feeling like that fact matters.

prakhar1144 replied:
Ah right, `AUTOMATICALLY_UNMUTE_TOPICS_ON_INITIATION`. The tests didn't fail because internally they have the same `integer` value.

@timabbott  Sorry, I'm unsure about what you meant by:
"use a common set of values, and name them e.g. CHANGE_USER_TOPIC_ON_INITIATION"

timabbott replied:
Oh I meant to just rename both `AUTOMATICALLY_UNMUTE_TOPICS_ON_INITIATION` and `AUTOMATICALLY_FOLLOW_TOPICS_ON_INITIATION` with `CHANGE_USER_TOPIC_ON_INITIATION` -- so we'd just have one set of names for the 1/2/3 values and it'd feel very clear in the code that they're supposed to be parallel.

prakhar1144 replied:
Alright, this is the modified code in `zerver/models.py` (removes duplication)

```python
    # The following two settings control which topics to automatically
    # 'follow' or 'unmute in a muted stream', respectively.
    # Follow or unmute a topic automatically on:
    # - PARTICIPATION: Send a message, React to a message, Participate in a poll or Edit a TO-DO list.
    # - INITIATION: Send the first message in the topic.
    # - NEVER: Never automatically follow or unmute a topic.
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_PARTICIPATION = 1
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION = 2
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER = 3
    AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_CHOICES = [
        AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_PARTICIPATION,
        AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_INITIATION,
        AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER,
    ]
    automatically_follow_topics_policy = models.PositiveSmallIntegerField(
        default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER
    )
    automatically_unmute_topics_in_muted_streams_policy = models.PositiveSmallIntegerField(
        default=AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_NEVER
    )
```

I have made small changes to your suggestion `CHANGE_USER_TOPIC_ON_INITIATION`
* used the term `VISIBILITY_POLICY` instead of `USER_TOPIC`
* prefixed with `AUTOMATICALLY_` for readability.

without the prefix, things like `CHANGE_USER_TOPIC_NEVER` seems to convey that "never change the visibility policy" instead of "never automatically change the visibility policy"

timabbott replied:
OK, works for me!

timabbott commented on zerver/tests/test_user_topics.py:
```diff
@@ -638,3 +641,300 @@ def test_unmuted_topic_remove_invalid(self) -> None:
+            UserProfile.AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION,
+            acting_user=None,
+        )
+        # For aaron, 'automatically_follow_topics_policy' NOT set to 'AUTOMATICALLY_FOLLOW_TOPICS_ON_PARTICIPATION'.
```
Yeah, that's probably OK.

timabbott reviewed (commented):
I posted a few more comments on the backend changes.

prakhar1144 commented:
The recent push includes a [small fix](https://github.com/zulip/zulip/pull/26054?notification_referrer_id=NT_kwDOA2JrwbM2ODA2MDk1MDQ5OjU2NzgxNzYx#discussion_r1327535603). + A couple of questions in the review comments above.

alya commented:
Let's try adding the note discussed on CZO just below the "Topic notifications" heading: "You will automatically follow topics that you have configured to both follow and unmute."

alya commented:
I also posted a suggestion [on CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1642367).

Otherwise looks good to me in manual testing! I have not attempted to test all the possible combinations of settings.

prakhar1144 commented:
Updates in the recent push:
* change the unmute option to "Automatically unmute topics in muted streams" -- [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1642367)
* Addressed the review comments
* Added a note in settings panel - [Alya's review comment](https://github.com/zulip/zulip/pull/26054#issuecomment-1724147122)

@alya I have added a demo video that shows:
* unmute policy only applicable in muted streams
* follow policy applicable in both the muted & not muted stream (as before)
* [Note added](https://github.com/zulip/zulip/pull/26054#issuecomment-1724147122) in settings panel

Let's fix the compose banner issue [alya mentioned](https://github.com/zulip/zulip/pull/26054#issuecomment-1724162776) in a separate follow-up PR? -- this PR already has ~85+ comments now.

timabbott commented on zerver/actions/reactions.py:
```diff
@@ -82,6 +89,31 @@ def do_add_reaction(
+        stream_id = message.recipient.type_id
+        (stream, sub) = access_stream_by_id(user_profile, stream_id)
+        assert stream is not None
+        if sub:
```
This might be better with the early return pattern.

timabbott replied:
Or maybe better is to extract this whole block as a function we can call; `user_topics.process_participation(user_profile, message)` as the part inside the `if` statement; and then this could be called both here and in the submessage code path?

timabbott replied:
This works for me!

timabbott commented on zerver/actions/submessage.py:
```diff
@@ -48,6 +51,29 @@ def do_add_submessage(
+    # Determine and set the visibility_policy depending on 'automatically_follow_topics_policy'
+    # and 'automatically_unmute_topics_policy'.
+    if submessage.message.is_stream_message():
+        stream_id = submessage.message.recipient.type_id
```
This function doesn't check the `is_bot` case that the reactions code path does; should it? I think maybe bots can't use submessages, but it might be more future-proof to to the exact check we have for reactions here.

prakhar1144 replied:
Used `set_visibility_policy_possible` as explained in the above comment.

timabbott commented:
> Let's fix the compose banner issue alya mentioned in a separate follow-up PR?

Yeah that sounds good to me, that should be a straightforward follow-up.

prakhar1144 commented:
Addressed the review comments (see respective comments):
* [ Important](https://github.com/zulip/zulip/pull/26054#discussion_r1337766088)

timabbott commented on zerver/lib/message.py:
```diff
@@ -1697,3 +1698,136 @@ def update_to_dict_cache(
+            return None
+        return visibility_policy
+
+    old_accessible_messages_in_topic = UserMessage.objects.filter(
```
Let's add a comment here "# Now we need to check if the user initiated the topic. We use the user's own message access to avoid leaking information in private streams with shared history."

prakhar1144 replied:
Done, see the below review comment.

timabbott commented on zerver/lib/message.py:
```diff
@@ -1697,3 +1698,136 @@ def update_to_dict_cache(
+        user_profile=sender,
+        message__recipient_id=message.recipient_id,
+        message__subject__iexact=message.topic_name(),
+    ).exclude(message_id=message.id)
```
Reading this check, I realized that I think it is wrong for in the case of a stream with shared history; in that case, 
the initiation check needs to be more like `get_topic_history_for_public_stream` not `get_topic_history_for_stream` -- basically we need a `can_access_stream_history` if/else clause here, and filter on just `Message` if the user has history access.

prakhar1144 replied:
Updated code:

```python
    # Now we need to check if the user initiated the topic.
    old_accessible_messages_in_topic: Union[QuerySet[Message], QuerySet[UserMessage]]
    if can_access_stream_history(sender, stream):
        old_accessible_messages_in_topic = messages_for_topic(
            realm_id=sender.realm_id,
            stream_recipient_id=message.recipient_id,
            topic_name=message.topic_name(),
        ).exclude(id=message.id)
    else:
        # We use the user's own message access to avoid leaking information in
        # private streams with protected history.
        old_accessible_messages_in_topic = UserMessage.objects.filter(
            user_profile=sender,
            message__recipient_id=message.recipient_id,
            message__subject__iexact=message.topic_name(),
        ).exclude(message_id=message.id)
```

timabbott commented:
I rebased this to resolve the merge conflicts; there's a test failure that I suspect is just because some new tests were written without 49092dfa79e10aa8fa28e01d6991e9bba065464c already in `main`. 

I think the above is the complete list of outstanding work to integrate this feature. However, I think we might want to take the opportunity to change the default to "PARTICIPATION" for "automatically unmute" and `INITIATION` for "automatically follow" -- and then test that this interacts nicely with our compose banners (E.g. we don't pop up a banner telling you to unmute a topic you sent to a muted stream if your configuration automatically unmuted it when you sent the message). 

Hrm, I kinda feel like "participation" might feel a bit aggressive for "automatically unmute" -- I'm now wondering if we want an intermediate "Send a message" that doesn't include reactions/etc. Will chat with @alya about that and some of the transition trade-offs today.

Maybe that can be done as a new commit intended to be squashed into the existing commit, just to keep those changes separate.

timabbott commented:
OK I chatted with @alya and I think there's a few changes:
* We should add a new `AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_SEND` ("Topics I send a message to") -- correct me if I'm wrong, but I think this is a pretty simple change. And set that as the default for `automatically_unmute_topics_in_muted_streams_policy`. (And `ON_INITIATE` is confirmed as the planned default for the follow topic setting)
* I opened https://github.com/zulip/zulip/issues/26900 as a follow-up for communicating more clearly about when we're making changes.

prakhar1144 commented:
Posted couple of blockers:
* [CZO - 1](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/register.3A.20.60unread_msgs.2Ementions.60/near/1649996)
* [CZO -2](https://chat.zulip.org/#narrow/stream/49-development-help/topic/message.20unread.20for.20sender/near/1649896)

The issue is when we send a message in muted stream and automatically unmute/follow a topic the `unread_mssg.count` should increase as it represents not muted messages count. Web client doesn't use `count`. `apply_events` should also handle such cases.

The tests are failing as we fixed the `count` to consider `unmuted/followed` topics in #26848

prakhar1144 commented:
[Posted a design question](https://chat.zulip.org/#narrow/stream/101-design/topic/automatically.20follow.20topics.20.2325914/near/1650898)

prakhar1144 commented on zerver/lib/message.py:
```diff
@@ -1765,25 +1785,37 @@ def visibility_policy_for_send_message(
 
-    visibility_policy = visibility_policy_for_participation(sender, is_stream_muted)
+    visibility_policy_participation = visibility_policy_for_participation(sender, is_stream_muted)
+    visibility_policy_send = visibility_policy_for_send(sender, is_stream_muted)
```
`visibility_policy_send` doesn't look a perfect name for this. Any better suggestions?

timabbott replied:
I don't have a readily available better idea; I think it'll be clear enough from the context. So let's just live with it and we can rethink when next revising that code path.

prakhar1144 commented:
Updates in the recent push:

* First commit to fix a bug
* Second commit to address the issue [discussed on CZO
](https://chat.zulip.org/#narrow/stream/49-development-help/topic/message.20unread.20for.20sender/near/1649896) -- This was causing a test failure, so included as a prep commit.
* Added the `AUTOMATICALLY_CHANGE_VISIBILITY_POLICY_ON_SEND` ("Topics I send a message to") option.
* Addressed the review comments

I think changing the defaults would make sense only when we will fix the compose banners flow in a follow-up PR?

timabbott commented on zerver/lib/events.py:
```diff
@@ -715,7 +715,7 @@ def apply_event(
     if event["type"] == "message":
         state["max_message_id"] = max(state["max_message_id"], event["message"]["id"])
-        if "raw_unread_msgs" in state:
+        if "raw_unread_msgs" in state and "read" not in event["flags"]:
```
This is a good bug fix for the production code! (Well, very unlikely case, but nice to have fixed).

I would have put this in its own commit so it gets more emphasis; I'll try extracting it.

timabbott replied:
Ended up deciding it's not worth the effort to extract, so just mentioned it in the commit message.

timabbott commented on zerver/models.py:
```diff
@@ -3123,6 +3123,7 @@ def sent_by_human(self) -> bool:
                 "website",
                 "ios",
                 "android",
+                "test suite",
```
Thinking about this a bit more, I think we should add this only if `settings.TEST_SUITE`; going to try doing that change for this commit.

timabbott commented on zerver/lib/message.py:
```diff
@@ -1110,6 +1110,7 @@ def extract_unread_data_from_um_rows(
         return raw_unread_messages
 
     muted_stream_ids = get_muted_stream_ids(user_profile)
+    raw_unread_messages["muted_stream_ids"] = muted_stream_ids
```
This bug fix would have been much better to submit an independent PR for, since it's fixing a regression that's in `main`; the bug in question went out to production last night because I hadn't had a chance to re-review this.

timabbott commented on zerver/models.py:
```diff
@@ -3111,20 +3137,26 @@ def sent_by_human(self) -> bool:
+            # Since the vast majority of messages are sent by humans
+            # in Zulip, treat test suite messages as such.
+            or (sending_client == "test suite" and settings.TEST_SUITE)
+        )
```
Applied this revision so that this logic would only apply in the test suite; it's best to avoid changing production outcomes for testing reasons even if it's probably not going to come up in practice:

```diff
diff --git a/zerver/models.py b/zerver/models.py
index 55e958d22f..10129cb14d 100644
--- a/zerver/models.py
+++ b/zerver/models.py
@@ -3137,21 +3137,26 @@ class Message(AbstractMessage):
         sending_client = self.sending_client.name.lower()
 
         return (
-            sending_client
-            in (
-                "zulipandroid",
-                "zulipios",
-                "zulipdesktop",
-                "zulipmobile",
-                "zulipelectron",
-                "zulipterminal",
-                "snipe",
-                "website",
-                "ios",
-                "android",
-                "test suite",
+            (
+                sending_client
+                in (
+                    "zulipandroid",
+                    "zulipios",
+                    "zulipdesktop",
+                    "zulipmobile",
+                    "zulipelectron",
+                    "zulipterminal",
+                    "snipe",
+                    "website",
+                    "ios",
+                    "android",
+                )
             )
-        ) or ("desktop app" in sending_client)
+            or ("desktop app" in sending_client)
+            # Since the vast majority of messages are sent by humans
+            # in Zulip, treat test suite messages as such.
+            or (sending_client == "test suite" and settings.TEST_SUITE)
+        )
 
     @staticmethod
     def is_status_message(content: str, rendered_content: str) -> bool:
```

timabbott commented:
This looks great! Marked to merge once CI passes, leaving the `never` default in place -- we will do a future migration to change the defaults of these settings once we have everything finished for the feature. 

Huge thanks for all the work on this @prakhar1144!
